File: packages/core/src/gpu/device.ts
import type { GPUContext } from "../types.js";

export class GPUDeviceError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GPUDeviceError";
  }
}

/**
 * Initialize WebGPU adapter and device.
 * Requests maximum buffer sizes and compute invocations for large model support.
 *
 * @param existingDevice - Optional existing GPU device to reuse.
 * @returns GPU context with device, adapter, and limits.
 * @throws {GPUDeviceError} If WebGPU is not available or adapter request fails.
 *
 * @example
 * ```ts
 * const gpu = await initGPU();
 * console.log("Max buffer size:", gpu.limits.maxBufferSize);
 * ```
 */
export async function initGPU(existingDevice?: GPUDevice): Promise<GPUContext> {
  if (existingDevice) {
    return {
      device: existingDevice,
      adapter: null,
      limits: existingDevice.limits,
    };
  }

  if (typeof navigator === "undefined" || !navigator.gpu) {
    throw new GPUDeviceError(
      "WebGPU is not supported in this environment. " +
        "Please use a browser with WebGPU support (Chrome 113+, Edge 113+, Firefox Nightly)."
    );
  }

  const adapter = await navigator.gpu.requestAdapter({
    powerPreference: "high-performance",
  });

  if (!adapter) {
    throw new GPUDeviceError(
      "Failed to obtain WebGPU adapter. Check that your GPU drivers are up to date."
    );
  }

  const requiredLimits: Record<string, number> = {};

  // Request max buffer sizes for large weight tensors
  requiredLimits.maxBufferSize = adapter.limits.maxBufferSize;
  requiredLimits.maxStorageBufferBindingSize = adapter.limits.maxStorageBufferBindingSize;

  // Request max buffer count for bind groups
  const maxBuffersPerGroup =
    adapter.limits.maxStorageBuffersPerShaderStage;
  requiredLimits.maxStorageBuffersPerShaderStage = maxBuffersPerGroup;

  // Request max compute workgroup sizes
  requiredLimits.maxComputeWorkgroupSizeX =
    adapter.limits.maxComputeWorkgroupSizeX;
  requiredLimits.maxComputeWorkgroupSizeY =
    adapter.limits.maxComputeWorkgroupSizeY;
  requiredLimits.maxComputeWorkgroupSizeZ =
    adapter.limits.maxComputeWorkgroupSizeZ;
  requiredLimits.maxComputeInvocationsPerWorkgroup =
    adapter.limits.maxComputeInvocationsPerWorkgroup;
  requiredLimits.maxComputeWorkgroupStorageSize =
    adapter.limits.maxComputeWorkgroupStorageSize;

  const device = await adapter.requestDevice({
    requiredLimits,
  });

  device.lost.then((info) => {
    console.error(`WebGPU device lost: ${info.message} (reason: ${info.reason})`);
  });

  return { device, adapter, limits: device.limits };
}

File: packages/rust/crates/oxbitnet/src/gpu/device.rs
use std::sync::Arc;

use crate::error::{BitNetError, Result};

pub struct GpuContext {
    pub device: Arc<wgpu::Device>,
    pub queue: Arc<wgpu::Queue>,
}

/// Initialize wgpu device and queue with maximum limits.
pub async fn init_gpu() -> Result<GpuContext> {
    let instance = wgpu::Instance::default();

    let adapter = instance
        .request_adapter(&wgpu::RequestAdapterOptions {
            power_preference: wgpu::PowerPreference::HighPerformance,
            ..Default::default()
        })
        .await
        .map_err(|e| BitNetError::Gpu(format!("Failed to get adapter: {e}")))?;

    let mut required_limits = wgpu::Limits::default();
    let adapter_limits = adapter.limits();

    required_limits.max_buffer_size = adapter_limits.max_buffer_size;
    required_limits.max_storage_buffer_binding_size =
        adapter_limits.max_storage_buffer_binding_size;
    required_limits.max_storage_buffers_per_shader_stage =
        adapter_limits.max_storage_buffers_per_shader_stage;
    required_limits.max_compute_workgroup_size_x =
        adapter_limits.max_compute_workgroup_size_x;
    required_limits.max_compute_workgroup_size_y =
        adapter_limits.max_compute_workgroup_size_y;
    required_limits.max_compute_workgroup_size_z =
        adapter_limits.max_compute_workgroup_size_z;
    required_limits.max_compute_invocations_per_workgroup =
        adapter_limits.max_compute_invocations_per_workgroup;
    required_limits.max_compute_workgroup_storage_size =
        adapter_limits.max_compute_workgroup_storage_size;

    let (device, queue) = adapter
        .request_device(&wgpu::DeviceDescriptor {
            label: Some("oxbitnet"),
            required_features: wgpu::Features::empty(),
            required_limits,
            ..Default::default()
        })
        .await?;

    Ok(GpuContext {
        device: Arc::new(device),
        queue: Arc::new(queue),
    })
}
